test/common/upstream/hds_test.cc

```diff
@@ -678,5 +693,473 @@ TEST_F(HdsTest, TestSendResponseOneEndpointTimeout) {
+            Network::MockClientConnection* connection =
+                new NiceMock<Network::MockClientConnection>();
+
+            // pretend our endpoint was connected to.
+            connection->raiseEvent(Network::ConnectionEvent::Connected);
+
+            // return this new, connected endpoint.
+            return connection;
+          }));
+
+  EXPECT_CALL(*server_response_timer_, enableTimer(_, _)).Times(AtLeast(1));
+  EXPECT_CALL(async_stream_, sendMessageRaw_(_, false));
+  EXPECT_CALL(test_factory_, createClusterInfo(_)).WillRepeatedly(Return(cluster_info_));
+  EXPECT_CALL(dispatcher_, deferredDelete_(_)).Times(AtLeast(1));
+  // Process message
+  hds_delegate_->onReceiveMessage(std::move(message));
+  hds_delegate_->sendResponse();
+
+  // Save list of health checkers for use later.
+  auto original_hcs = hds_delegate_->hdsClusters()[0]->healthCheckers();
+  ASSERT_EQ(original_hcs.size(), 2);
+
+  // Create a new specifier, but make the second health checker different and add a third.
+  // Then reverse the order so the first one is at the end, testing the hashing works as expected.
+  message.reset(createSimpleMessage());
+  auto new_hc0 = message->mutable_cluster_health_checks(0)->add_health_checks();
+  new_hc0->MergeFrom(message->mutable_cluster_health_checks(0)->health_checks(0));
+  new_hc0->mutable_http_health_check()->set_path("/path0");
+  auto new_hc1 = message->mutable_cluster_health_checks(0)->add_health_checks();
+  new_hc1->MergeFrom(message->mutable_cluster_health_checks(0)->health_checks(0));
+  new_hc1->mutable_http_health_check()->set_path("/path1");
+  message->mutable_cluster_health_checks(0)->mutable_health_checks()->SwapElements(0, 2);
+  hds_delegate_->onReceiveMessage(std::move(message));
+
+  // Get the new health check list from HDS.
+  auto new_hcs = hds_delegate_->hdsClusters()[0]->healthCheckers();
+  ASSERT_EQ(new_hcs.size(), 3);
+
+  // Make sure our first hc from the original list is the same as the third in the new list.
+  EXPECT_EQ(original_hcs[0], new_hcs[2]);
+  EXPECT_TRUE(original_hcs[1] != new_hcs[1]);
+
+  // Check to see that HDS got two requests, and updated two times with it.
+  checkHdsCounters(2, 0, 0, 2);
+}
+
+// Test to see that if clusters with an empty name get used, there are two clusters.
+// Also test to see that if two clusters with the same non-empty name are used, only have
+// One cluster.
+TEST_F(HdsTest, TestClusterSameName) {
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  EXPECT_CALL(async_stream_, sendMessageRaw_(_, _));
+  createHdsDelegate();
+
+  // Create Message
+  message = createComplexSpecifier(2, 1, 1);
+  // Set both clusters to have an empty name.
+  message->mutable_cluster_health_checks(0)->set_cluster_name("");
+  message->mutable_cluster_health_checks(1)->set_cluster_name("");
+
+  // Create a new active connection on request, setting its status to connected
```

Comment:
Nit: I think this could be factored out into a `expecCreateClientConnection()` method in the base class test fixture, it might make tests a bit less verbose. Also consider other possible boiler plate reductions that don't make the test harder to read.

Reply:
I agree, especially since I use this expanded version many times. Changed.